WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorViewModel.swift

```diff
@@ -0,0 +1,84 @@
+import Foundation
+
+final class CustomFieldEditorViewModel: ObservableObject {
+    @Published var key: String {
+        didSet {
+            validateKey(key)
+        }
+    }
+
+    @Published var value: String
+    @Published private(set) var keyErrorMessage: String?
+
+    /// To be set by the View
+    @Published var notice: Notice?
+
+    private let initialKey: String
+    private let initialValue: String
+
+    private let disallowedKeys: [String]
+
+    private let onSave: (String, String) -> Void
+
+    /// Closure invoked during field deletion. Optional as it's not needed when creating a new field. Not private because it is used in the view for display logic.
+    let onDelete: (() -> Void)?
+
+    var hasUnsavedChanges: Bool {
+        key != initialKey || value != initialValue
+    }
+
+    /// Note that `isEmpty` is validated separately, because we don't want to show error message when it's still empty, and instead just disable the button.
+    var hasValidKey: Bool {
+        !key.isEmpty && keyErrorMessage == nil
+    }
+
+    init(key: String,
+         value: String,
+         disallowedKeys: [String] = [],
+         onSave: @escaping (String, String) -> Void,
+         onDelete: (() -> Void)? = nil) {
+        self.key = key
+        self.value = value
+        self.initialKey = key
+        self.initialValue = value
+        self.disallowedKeys = disallowedKeys
+        self.onSave = onSave
+        self.onDelete = onDelete
+    }
+
+    func validateKey(_ newValue: String) {
+        if newValue.hasPrefix("_") {
+            keyErrorMessage = Localization.keyErrorPrefix
+        } else if disallowedKeys.contains(newValue) {
+            keyErrorMessage = Localization.keyErrorDisallowedKey
```

Comment:
Thanks for sharing this!

> To prevent the error message from appearing multiple times while the user is still editing, which is likely when they haven't finished.

I did see the validation for duplicates is done on tap while testing and checking Android's error message, but did not know the reason. That's a good point also because now iOS does a comparison check for every single character typed.

> We fetch the duplicate keys from DB (as you know, passing large arguments in Android is not recommended)

I got curious so I tested by generating 500 random keys, each having 12 characters, then passing them. Anecdotally, from observing CPU and memory in the Debug Navigator, the CPU jumps a bit each time I'm typing, but goes down quickly when I stopped typing. The memory stayed around 170-180 MB similarly, both when using these 500 random keys or 1 real key.

I replaced the comparison from using `.contains` to `.first` to optimize a little bit more https://github.com/woocommerce/woocommerce-ios/pull/14235/commits/3566d683c7b5a5a72486e33f847a526de434ef11 

Given that it's a bit unlikely that an Order or Product contains 500 custom fields, and with the result above, I suppose things are fine as-is now. We can refactor later if it's causing extreme slowdown.